regression_em/model.py

```diff
@@ -1,18 +1,287 @@
+"""Package to perform regression EM algorithm self.
+For the detail of this algortithm,
+["Position Bias Estimation for Unbiased Learning to Rank in Personal Search" Xuanhui Wang et al.]
+(https://static.googleusercontent.com/media/research.google.com/ja//pubs/archive/46485.pdf)
+"""
+from random import random
+# from typing import Sequence, Tuple, Union
+from typing import Tuple, Union
+
+import numpy as np
 import scipy.sparse as sp
-from sklearn.linear_model import LinearRegression
+from sklearn.linear_model import LinearRegression, Ridge
+from sklearn.base import BaseEstimator
+
+
+# Define type.
+Matrix = Union[np.ndarray, sp.csr_matrix, sp.csc_matrix]
+RegParam = Tuple[Matrix, float]
+
+
+class RegressionEM(BaseEstimator):
+    """Regression EM can estimate latent factors based on below model.
+
+       Outcome = Left latent * Right latent
+
+       For the detail of Regression EM algortithm,
+       ["Position Bias Estimation for Unbiased Learning to Rank in Personal Search" Xuanhui Wang et al.]
+       (https://static.googleusercontent.com/media/research.google.com/ja//pubs/archive/46485.pdf)
+
+       :Parameters:
+
+       alpha : float, default=1
+       Weights associated with regularization term
+
+       max_iter : int, default=100
+       Maximum number of iterations taken for the solvers to converge
+
+       epsilon : float, default=1e-10
+       Tiny value to avoid division by zero.
+
+       class_weights : 'balanced', default = None
+       The "balanced" mode uses the values of y to automatically adjust
+       weights inversely proportional to class frequencies in the input data
+       as ``n_samples / (n_classes * np.bincount(y))``
+       Note that these weights will be multiplied with sample_weight (passed
+       through the fit method) if sample_weight is specified.
+
+       feature_index : int
+       the index to devide X into left latent features and right latent features.
+
+       :Example:
+
+       >>> import RegressionEM
+       >>> left_feat = array([[0.17843638, 0.09311004, 0.89600447, ...],
+       [0.55349066, 0.83427622, 0.34841103, ...],
+       ...,
+       [0.22199485, 0.19540406, 0.02678277, ...],
+       [0.62612729, 0.71996384, 0.66445362, ...]])
+       >>> right_feat = array([[0.6177678 , 0.69322733, 0.95146727, ...],
+       [0.96681348, 0.79037145, 0.45834361, ...],
+       ...,
+       [0.64773992, 0.86541352, 0.04755084, ...],
+       [0.37910497, 0.44344932, 0.48168189, ...]])
+       >>> X = np.hstack([left_feat, right_feat])
+       >>> y = array([0, 0, 0, ...,  1, 0, 1])
+       >>> rem = RegressionEM(max_iter=100, with_sample_weights=True, alpha=1)
+       >>> rem.fit(X, y, 100)
+    """
+
+    def __init__(self, alpha: float = 0, max_iter: int = 100, epsilon: float = 10 ** -10, class_weights: str = None, split_index: int = 0) -> None:
```

Comment:
split_index or feature_index?

Reply:
Do not give a default value to `split_index` because this argument must be initialized.